csharp/extractor/Semmle.Extraction/TrapExtensions.cs

```diff
@@ -0,0 +1,250 @@
+        }
+
+        public static void WriteTrapString(this TextWriter trapFile, string s)
+        {
+            trapFile.Write('\"');
+            if (NeedsTruncation(s))
+            {
+                // Slow path
+                int remaining = maxStringBytes;
+                WriteTruncatedString(trapFile, s, ref remaining);
+            }
+            else
+            {
+                // Fast path
+                WriteString(trapFile, s);
+            }
+            trapFile.Write('\"');
+        }
+
+        public static void WriteTrapFloat(this TextWriter trapFile, float f)
+        {
+            trapFile.Write(f.ToString("0.#####e0"));  // Trap importer won't accept ints
+        }
+
+        public static FirstParam BeginTuple(this TextWriter trapFile, string name)
+        {
+            trapFile.Write(name);
+            trapFile.Write('(');
+            return new FirstParam(trapFile);
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1)
+        {
+            trapFile.BeginTuple(name).Param(p1).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, IEntity p2)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, string p2, IEntity p3, IEntity p4)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).Param(p4).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, string p2, IEntity p3)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, int p2, IEntity p3)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, int p2, int p3)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, IEntity p2, int p3)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, IEntity p2, IEntity p3)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, string p2, IEntity p3, IEntity p4, IEntity p5)
+        {
+            trapFile.BeginTuple(name).Param(p1).Param(p2).Param(p3).Param(p4).Param(p5).EndTuple();
+        }
+
+        public static void WriteTuple(this TextWriter trapFile, string name, IEntity p1, int p2)
```

Comment:
I've simplified this just to use `object[]` plus a few specialisations for efficiency. The intention was to avoid the switching on types, which is ugly and potentially slow, but the resulting code was quite clunky as you correctly pointed out. There doesn't need to be a measurable impact of this so I went with the simpler approach.